source/extensions/filters/http/ext_proc/processor_state.cc

```diff
@@ -67,44 +68,116 @@ bool ProcessorState::handleHeadersResponse(const HeadersResponse& response) {
+          clearWatermark();
+          return true;
+        }
+
+        if (hasBufferedData()) {
+          // We now know that we need to process what we have buffered in streaming mode.
+          // Move the current buffer into the queue for remote processing and clear the
+          // buffered data.
+          Buffer::OwnedImpl buffered_chunk;
+          modifyBufferedData(
+              [&buffered_chunk](Buffer::Instance& data) { buffered_chunk.move(data); });
+          ENVOY_LOG(debug, "Sending first chunk using buffered data");
+          filter_.sendBodyChunk(*this, buffered_chunk,
+                                ProcessorState::CallbackState::StreamedBodyCallback, false);
+          enqueueStreamingChunk(buffered_chunk, false, true);
+          if (!queueOverBufferLimit()) {
+            clearWatermark();
+          }
+        } else {
+          clearWatermark();
+        }
+        continueIfNecessary();
         return true;
       }
 
       if (send_trailers_ && trailers_available_) {
         // Trailers came in while we were waiting for this response, and the server
         // is not interested in the body, so send them now.
         filter_.sendTrailers(*this, *trailers_);
+        clearWatermark();
         return true;
       }
     }
 
     // If we got here, then the processor doesn't care about the body or is not ready for
     // trailers, so we can just continue.
     headers_ = nullptr;
-    continueProcessing();
+    continueIfNecessary();
+    clearWatermark();
     return true;
   }
   return false;
 }
 
 bool ProcessorState::handleBodyResponse(const BodyResponse& response) {
-  if (callback_state_ == CallbackState::BufferedBodyCallback) {
-    ENVOY_LOG(debug, "Applying body response to buffered data");
-    modifyBufferedData([this, &response](Buffer::Instance& data) {
-      MutationUtils::applyCommonBodyResponse(response, headers_, data);
-    });
+  bool should_continue = false;
+  if (callback_state_ == CallbackState::BufferedBodyCallback ||
+      callback_state_ == CallbackState::StreamedBodyCallback ||
+      callback_state_ == CallbackState::StreamedBodyCallbackFinishing) {
+    ENVOY_LOG(debug, "Processing body response");
+    if (callback_state_ == CallbackState::BufferedBodyCallback) {
+      ENVOY_LOG(debug, "Applying body response to buffered data. State = {}", callback_state_);
+      modifyBufferedData([this, &response](Buffer::Instance& data) {
+        MutationUtils::applyCommonBodyResponse(response, headers_, data);
+      });
+      clearWatermark();
+      callback_state_ = CallbackState::Idle;
+      should_continue = true;
+
+    } else if (callback_state_ == CallbackState::StreamedBodyCallback ||
+               callback_state_ == CallbackState::StreamedBodyCallbackFinishing) {
+      bool delivered_one = false;
+      while (auto queued_chunk = dequeueStreamingChunk(delivered_one)) {
+        // Loop through queue in case some of it is chunks that were never
```

Comment:
It's a little unclear to me when streaming can degenerate into general buffering. I think above, before headers are received back from the ext_proc server, this can happen. Here we dequeue chunks, but this is subject to just the stream buffer that we are watermarking?

Reply:
Streaming degenerates into buffering if, while we were waiting for the response to the "headers" message from the external processor, the entire message body comes in to the filter and is buffered. At that point, whether the processor requests "streamed" or "buffered" mode it's the same thing since we already have the whole message in the buffer.